### DISCUSSION TITLE
Add a `consume` paramter to request operations

### DISCUSSION THREAD

**Discussion by adriangb** [encode/starlette#1520] [Ideas]
Currently if you want to parse a form without consuming the body you have to do:

```python
await request.body()  # caches the body to memory
await request.form()  # parses the cached body as a form
```

It might be nice to add a `Request.form(consume=False)` API and a `Request.json(consume=True)` API to make this more intuitive.
Also for the JSON case there is no way to read a JSON body without duplicating it in memory, although JSON parsing is much easier than form parsing so it's not _that_ big of a deal (it would be if it was the other way around).

**Comment by Kludex** [comment]
I'll close this as outdated. Correct me if wrong.

**Reply by adriangb** [reply]
I forgot about this. I still think it’s relevant. There’s been some requests for form not consuming the body which I’ve lost track of.
